[multistage][bug] literal attachment not working properly with other expansion rules

see:
```
SELECT ,
  AVG(int_val) AS average,
  PERCENTILE(int_val, 50) AS median
FROM
  tbl
```
will cause 
```
Error Code: 200

java.util.concurrent.ExecutionException: org.apache.pinot.spi.exception.BadQueryRequestException: Invalid aggregation function: percentile($0,$1); Reason: null
	at java.base/java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:395)
	at java.base/java.util.concurrent.CompletableFuture.get(CompletableFuture.java:2022)
	at org.apache.pinot.query.service.server.QueryServer.submit(QueryServer.java:126)
	at org.apache.pinot.common.proto.PinotQueryWorkerGrpc$MethodHandlers.invoke(PinotQueryWorkerGrpc.java:286)
...
Caused by: org.apache.pinot.spi.exception.BadQueryRequestException: Invalid aggregation function: percentile($0,$1); Reason: null
	at org.apache.pinot.core.query.aggregation.function.AggregationFunctionFactory.getAggregationFunction(AggregationFunctionFactory.java:372)
	at org.apache.pinot.query.runtime.operator.AggregateOperator.getAggFunctionForRawInput(AggregateOperator.java:250)
	at org.apache.pinot.query.runtime.operator.AggregateOperator.getAggFunctions(AggregateOperator.java:205)
	at org.apache.pinot.query.runtime.operator.AggregateOperator.<init>(AggregateOperator.java:97)
org.apache.pinot.query.service.dispatch.QueryDispatcher.submit(QueryDispatcher.java:148)
org.apache.pinot.query.service.dispatch.QueryDispatcher.submitAndReduce(QueryDispatcher.java:92)
org.apache.pinot.broker.requesthandler.MultiStageBrokerRequestHandler.handleRequest(MultiStageBrokerRequestHandler.java:188)
org.apache.pinot.broker.requesthandler.BaseBrokerRequestHandler.handleRequest(BaseBrokerRequestHandler.java:279)
```

--- a/pinot-query-planner/src/main/java/org/apache/calcite/rel/rules/PinotAggregateLiteralAttachmentRule.java
+++ b/pinot-query-planner/src/main/java/org/apache/calcite/rel/rules/PinotAggregateLiteralAttachmentRule.java
@@ -24,6 +24,7 @@ import java.util.List;
 import java.util.Map;
 import org.apache.calcite.plan.RelOptRule;
 import org.apache.calcite.plan.RelOptRuleCall;
+import org.apache.calcite.rel.RelNode;
 import org.apache.calcite.rel.core.Aggregate;
 import org.apache.calcite.rel.core.AggregateCall;
 import org.apache.calcite.rel.core.Project;
@@ -31,7 +32,6 @@ import org.apache.calcite.rel.hint.PinotHintOptions;
 import org.apache.calcite.rel.hint.PinotHintStrategyTable;
 import org.apache.calcite.rel.hint.RelHint;
 import org.apache.calcite.rel.logical.LogicalAggregate;
-import org.apache.calcite.rel.logical.LogicalProject;
 import org.apache.calcite.rex.RexLiteral;
 import org.apache.calcite.rex.RexNode;
 import org.apache.calcite.tools.RelBuilderFactory;
@@ -50,7 +50,7 @@ public class PinotAggregateLiteralAttachmentRule extends RelOptRule {
       new PinotAggregateLiteralAttachmentRule(PinotRuleUtils.PINOT_REL_FACTORY);
 
   public PinotAggregateLiteralAttachmentRule(RelBuilderFactory factory) {
-    super(operand(LogicalAggregate.class, some(operand(LogicalProject.class, any()))), factory, null);
+    super(operand(LogicalAggregate.class, any()), factory, null);
   }
 
   @Override
@@ -81,8 +81,8 @@ public class PinotAggregateLiteralAttachmentRule extends RelOptRule {
 
   private static Map<Pair<Integer, Integer>, RexExpression.Literal> extractLiterals(RelOptRuleCall call) {
     Aggregate aggregate = call.rel(0);
-    Project project = call.rel(1);
-    List<RexNode> rexNodes = project.getProjects();
+    RelNode input = PinotRuleUtils.unboxRel(aggregate.getInput());
+    List<RexNode> rexNodes = (input instanceof Project) ? ((Project) input).getProjects() : null;
     List<AggregateCall> aggCallList = aggregate.getAggCallList();
     final Map<Pair<Integer, Integer>, RexExpression.Literal> rexLiteralMap = new HashMap<>();
     for (int aggIdx = 0; aggIdx < aggCallList.size(); aggIdx++) {
@@ -93,9 +93,11 @@ public class PinotAggregateLiteralAttachmentRule extends RelOptRule {
         rexLiteralMap.put(new Pair<>(aggIdx, -1), new RexExpression.Literal(ColumnDataType.INT, argSize));
         // put the literals in to the map.
         for (int argIdx = 0; argIdx < argSize; argIdx++) {
-          RexNode field = rexNodes.get(aggCall.getArgList().get(argIdx));
-          if (field instanceof RexLiteral) {
-            rexLiteralMap.put(new Pair<>(aggIdx, argIdx), RexExpressionUtils.fromRexLiteral((RexLiteral) field));
+          if (rexNodes != null) {
+            RexNode field = rexNodes.get(aggCall.getArgList().get(argIdx));
+            if (field instanceof RexLiteral) {
+              rexLiteralMap.put(new Pair<>(aggIdx, argIdx), RexExpressionUtils.fromRexLiteral((RexLiteral) field));
+            }
           }
         }
       }

--- a/pinot-query-planner/src/main/java/org/apache/calcite/rel/rules/PinotQueryRuleSets.java
+++ b/pinot-query-planner/src/main/java/org/apache/calcite/rel/rules/PinotQueryRuleSets.java
@@ -110,11 +110,10 @@ public class PinotQueryRuleSets {
       PruneEmptyRules.UNION_INSTANCE
   );
 
-  // Pinot specific rules that should be run BEFORE all other rules
-  public static final Collection<RelOptRule> PINOT_PRE_RULES = ImmutableList.of(
+  // Pinot specific rules to run using a single RuleCollection since we attach aggregate info after optimizer.
+  public static final Collection<RelOptRule> PINOT_AGG_PROCESS_RULES = ImmutableList.of(
       PinotAggregateLiteralAttachmentRule.INSTANCE
   );
-
 
   // Pinot specific rules that should be run AFTER all other rules
   public static final Collection<RelOptRule> PINOT_POST_RULES = ImmutableList.of(

--- a/pinot-query-planner/src/main/java/org/apache/pinot/query/QueryEnvironment.java
+++ b/pinot-query-planner/src/main/java/org/apache/pinot/query/QueryEnvironment.java
@@ -128,15 +128,15 @@ public class QueryEnvironment {
     hepProgramBuilder.addMatchOrder(HepMatchOrder.DEPTH_FIRST);
 
     // ----
-    // Run Pinot specific pre-rules
-    hepProgramBuilder.addRuleCollection(PinotQueryRuleSets.PINOT_PRE_RULES);
-
-    // ----
     // Run the Calcite CORE rules using 1 HepInstruction per rule. We use 1 HepInstruction per rule for simplicity:
     // the rules used here can rest assured that they are the only ones evaluated in a dedicated graph-traversal.
     for (RelOptRule relOptRule : PinotQueryRuleSets.BASIC_RULES) {
       hepProgramBuilder.addRuleInstance(relOptRule);
     }
+
+    // ----
+    // Run Pinot rule to attach aggregation auxiliary info
+    hepProgramBuilder.addRuleCollection(PinotQueryRuleSets.PINOT_AGG_PROCESS_RULES);
 
     // ----
     // Pushdown filters using a single HepInstruction.

--- a/pinot-query-runtime/src/test/resources/queries/StatisticAggregates.json
+++ b/pinot-query-runtime/src/test/resources/queries/StatisticAggregates.json
@@ -144,16 +144,16 @@
     },
     "queries": [
       {
-        "sql": "SELECT groupingCol, COVARPOP(val, val), COVARSAMP(val, val), VARPOP(val), VARSAMP(val), STDDEVPOP(val), STDDEVSAMP(val) FROM {tbl} GROUP BY groupingCol",
-        "h2Sql": "SELECT groupingCol, COVAR_POP(val, val), COVAR_SAMP(val, val), VAR_POP(val), VAR_SAMP(val), STDDEV_POP(val), STDDEV_SAMP(val) FROM {tbl} GROUP BY groupingCol"
+        "sql": "SELECT groupingCol, AVG(val), COVARPOP(val, val), COVARSAMP(val, val), VARPOP(val), VARSAMP(val), STDDEVPOP(val), STDDEVSAMP(val) FROM {tbl} GROUP BY groupingCol",
+        "h2Sql": "SELECT groupingCol, AVG(val), COVAR_POP(val, val), COVAR_SAMP(val, val), VAR_POP(val), VAR_SAMP(val), STDDEV_POP(val), STDDEV_SAMP(val) FROM {tbl} GROUP BY groupingCol"
       },
       {
-        "sql": "SELECT COVARPOP(val, val), COVARSAMP(val, val), VARPOP(val), VARSAMP(val), STDDEVPOP(val), STDDEVSAMP(val) FROM {tbl} WHERE groupingCol='a'",
-        "h2Sql": "SELECT COVAR_POP(val, val), COVAR_SAMP(val, val), VAR_POP(val), VAR_SAMP(val), STDDEV_POP(val), STDDEV_SAMP(val) FROM {tbl} WHERE groupingCol='a'"
+        "sql": "SELECT AVG(val), COVARPOP(val, val), COVARSAMP(val, val), VARPOP(val), VARSAMP(val), STDDEVPOP(val), STDDEVSAMP(val) FROM {tbl} WHERE groupingCol='a'",
+        "h2Sql": "SELECT AVG(val), COVAR_POP(val, val), COVAR_SAMP(val, val), VAR_POP(val), VAR_SAMP(val), STDDEV_POP(val), STDDEV_SAMP(val) FROM {tbl} WHERE groupingCol='a'"
       },
       {
-        "sql": "SELECT t1.groupingCol, COVARPOP(t1.val, t2.val), COVARSAMP(t1.val, t2.val), VARPOP(t1.val + t2.val), VARSAMP(t1.val + t2.val), STDDEVPOP(t1.val + t2.val), STDDEVSAMP(t1.val + t2.val) FROM {tbl} AS t1 LEFT JOIN {tbl2} AS t2 USING (partitionCol) GROUP BY t1.groupingCol",
-        "h2Sql": "SELECT t1.groupingCol, COVAR_POP(t1.val, t2.val), COVAR_SAMP(t1.val, t2.val), VAR_POP(t1.val + t2.val), VAR_SAMP(t1.val + t2.val), STDDEV_POP(t1.val + t2.val), STDDEV_SAMP(t1.val + t2.val) FROM {tbl} AS t1 LEFT JOIN {tbl2} AS t2 USING (partitionCol) GROUP BY t1.groupingCol"
+        "sql": "SELECT t1.groupingCol, AVG(t1.val), COVARPOP(t1.val, t2.val), COVARSAMP(t1.val, t2.val), VARPOP(t1.val + t2.val), VARSAMP(t1.val + t2.val), STDDEVPOP(t1.val + t2.val), STDDEVSAMP(t1.val + t2.val) FROM {tbl} AS t1 LEFT JOIN {tbl2} AS t2 USING (partitionCol) GROUP BY t1.groupingCol",
+        "h2Sql": "SELECT t1.groupingCol, AVG(t1.val), COVAR_POP(t1.val, t2.val), COVAR_SAMP(t1.val, t2.val), VAR_POP(t1.val + t2.val), VAR_SAMP(t1.val + t2.val), STDDEV_POP(t1.val + t2.val), STDDEV_SAMP(t1.val + t2.val) FROM {tbl} AS t1 LEFT JOIN {tbl2} AS t2 USING (partitionCol) GROUP BY t1.groupingCol"
       },
       {
         "ignored": true,

--- a/pinot-query-runtime/src/test/resources/queries/UDFAggregates.json
+++ b/pinot-query-runtime/src/test/resources/queries/UDFAggregates.json
@@ -125,16 +125,16 @@
     },
     "queries": [
       {
-        "sql": "SELECT PERCENTILE(float_col, 50), PERCENTILE(double_col, 5), PERCENTILE(int_col, 75), PERCENTILE(long_col, 75) FROM {tbl}",
-        "outputs": [[1.75, 1, 150, 150]]
+        "sql": "SELECT AVG(int_col), PERCENTILE(float_col, 50), PERCENTILE(double_col, 5), PERCENTILE(int_col, 75), PERCENTILE(long_col, 75) FROM {tbl}",
+        "outputs": [[76.14285714285714, 1.75, 1, 150, 150]]
       },
       {
-        "sql": "SELECT bool_col, PERCENTILE(float_col, 50), PERCENTILE(double_col, 5), PERCENTILE(int_col, 75), PERCENTILE(long_col, 75) FROM {tbl} GROUP BY bool_col",
-        "outputs": [[false, 1.5, 1, 150, 150], [true, 300, 1.75, 175, 175]]
+        "sql": "SELECT bool_col, AVG(int_col), PERCENTILE(float_col, 50), PERCENTILE(double_col, 5), PERCENTILE(int_col, 75), PERCENTILE(long_col, 75) FROM {tbl} GROUP BY bool_col",
+        "outputs": [[false, 88.5, 1.5, 1, 150, 150], [true, 59.666666666666664, 300, 1.75, 175, 175]]
       },
       {
-        "sql": "SELECT /*+ aggOptions(is_skip_leaf_stage_aggregate='true') */ string_col, PERCENTILE(float_col, 50), PERCENTILE(double_col, 5), PERCENTILE(int_col, 75), PERCENTILE(long_col, 75) FROM {tbl} GROUP BY string_col",
-        "outputs": [["a", 400, 300, 2, 2], ["b", 100, 1, 100, 100], ["c", 1.5, 1.01, 175, 175]]
+        "sql": "SELECT /*+ aggOptions(is_skip_leaf_stage_aggregate='true') */ string_col, AVG(int_col), PERCENTILE(float_col, 50), PERCENTILE(double_col, 5), PERCENTILE(int_col, 75), PERCENTILE(long_col, 75) FROM {tbl} GROUP BY string_col",
+        "outputs": [["a", 2.0, 400, 300, 2, 2], ["b", 51.5, 100, 1, 100, 100], ["c", 142.0, 1.5, 1.01, 175, 175]]
       },
       {
         "sql": "SELECT PERCENTILE_EST(float_col, 50), PERCENTILE_EST(double_col, 5), PERCENTILE_EST(int_col, 75), PERCENTILE_EST(long_col, 75) FROM {tbl}",
@@ -201,16 +201,16 @@
     },
     "queries": [
       {
-        "sql": "select GET_THETA_SKETCH_ESTIMATE(DISTINCT_COUNT_RAW_THETA_SKETCH(string_col, 'nominalEntries=16')), GET_THETA_SKETCH_ESTIMATE(THETA_SKETCH_DIFF(DISTINCT_COUNT_RAW_THETA_SKETCH(int_col, ''), DISTINCT_COUNT_RAW_THETA_SKETCH(int_col, ''))), GET_THETA_SKETCH_ESTIMATE(THETA_SKETCH_UNION(DISTINCT_COUNT_RAW_THETA_SKETCH(int_col, ''), DISTINCT_COUNT_RAW_THETA_SKETCH(long_col, ''))), GET_THETA_SKETCH_ESTIMATE(THETA_SKETCH_INTERSECT(DISTINCT_COUNT_RAW_THETA_SKETCH(double_col, ''), DISTINCT_COUNT_RAW_THETA_SKETCH(float_col, 'nominalEntries=1000'))) FROM {tbl}",
-        "outputs": [[3, 0, 6, 6]]
+        "sql": "select AVG(int_col), GET_THETA_SKETCH_ESTIMATE(DISTINCT_COUNT_RAW_THETA_SKETCH(string_col, 'nominalEntries=16')), GET_THETA_SKETCH_ESTIMATE(THETA_SKETCH_DIFF(DISTINCT_COUNT_RAW_THETA_SKETCH(int_col, ''), DISTINCT_COUNT_RAW_THETA_SKETCH(int_col, ''))), GET_THETA_SKETCH_ESTIMATE(THETA_SKETCH_UNION(DISTINCT_COUNT_RAW_THETA_SKETCH(int_col, ''), DISTINCT_COUNT_RAW_THETA_SKETCH(long_col, ''))), GET_THETA_SKETCH_ESTIMATE(THETA_SKETCH_INTERSECT(DISTINCT_COUNT_RAW_THETA_SKETCH(double_col, ''), DISTINCT_COUNT_RAW_THETA_SKETCH(float_col, 'nominalEntries=1000'))) FROM {tbl}",
+        "outputs": [[76.14285714285714, 3, 0, 6, 6]]
       },
       {
-        "sql": "select bool_col, GET_THETA_SKETCH_ESTIMATE(DISTINCT_COUNT_RAW_THETA_SKETCH(string_col, 'nominalEntries=16')), GET_THETA_SKETCH_ESTIMATE(THETA_SKETCH_DIFF(DISTINCT_COUNT_RAW_THETA_SKETCH(int_col, ''), DISTINCT_COUNT_RAW_THETA_SKETCH(int_col, ''))), GET_THETA_SKETCH_ESTIMATE(THETA_SKETCH_UNION(DISTINCT_COUNT_RAW_THETA_SKETCH(int_col, ''), DISTINCT_COUNT_RAW_THETA_SKETCH(long_col, ''))), GET_THETA_SKETCH_ESTIMATE(THETA_SKETCH_INTERSECT(DISTINCT_COUNT_RAW_THETA_SKETCH(double_col, ''), DISTINCT_COUNT_RAW_THETA_SKETCH(float_col, 'nominalEntries=1000'))) FROM {tbl} GROUP BY bool_col",
-        "outputs": [[true, 2, 0, 2, 3], [false, 2, 0, 4, 3]]
+        "sql": "select bool_col, AVG(int_col), GET_THETA_SKETCH_ESTIMATE(DISTINCT_COUNT_RAW_THETA_SKETCH(string_col, 'nominalEntries=16')), GET_THETA_SKETCH_ESTIMATE(THETA_SKETCH_DIFF(DISTINCT_COUNT_RAW_THETA_SKETCH(int_col, ''), DISTINCT_COUNT_RAW_THETA_SKETCH(int_col, ''))), GET_THETA_SKETCH_ESTIMATE(THETA_SKETCH_UNION(DISTINCT_COUNT_RAW_THETA_SKETCH(int_col, ''), DISTINCT_COUNT_RAW_THETA_SKETCH(long_col, ''))), GET_THETA_SKETCH_ESTIMATE(THETA_SKETCH_INTERSECT(DISTINCT_COUNT_RAW_THETA_SKETCH(double_col, ''), DISTINCT_COUNT_RAW_THETA_SKETCH(float_col, 'nominalEntries=1000'))) FROM {tbl} GROUP BY bool_col",
+        "outputs": [[true, 59.666666666666664, 2, 0, 2, 3], [false, 88.5, 2, 0, 4, 3]]
       },
       {
-        "sql": "select /*+ aggOptions(is_skip_leaf_stage_aggregate='true') */ bool_col, GET_THETA_SKETCH_ESTIMATE(DISTINCT_COUNT_RAW_THETA_SKETCH(string_col, 'nominalEntries=16')), GET_THETA_SKETCH_ESTIMATE(THETA_SKETCH_DIFF(DISTINCT_COUNT_RAW_THETA_SKETCH(int_col, ''), DISTINCT_COUNT_RAW_THETA_SKETCH(int_col, ''))), GET_THETA_SKETCH_ESTIMATE(THETA_SKETCH_UNION(DISTINCT_COUNT_RAW_THETA_SKETCH(int_col, ''), DISTINCT_COUNT_RAW_THETA_SKETCH(long_col, ''))), GET_THETA_SKETCH_ESTIMATE(THETA_SKETCH_INTERSECT(DISTINCT_COUNT_RAW_THETA_SKETCH(double_col, ''), DISTINCT_COUNT_RAW_THETA_SKETCH(float_col, 'nominalEntries=1000'))) FROM {tbl} GROUP BY bool_col",
-        "outputs": [[true, 2, 0, 2, 3], [false, 2, 0, 4, 3]]
+        "sql": "select /*+ aggOptions(is_skip_leaf_stage_aggregate='true') */ bool_col, AVG(int_col), GET_THETA_SKETCH_ESTIMATE(DISTINCT_COUNT_RAW_THETA_SKETCH(string_col, 'nominalEntries=16')), GET_THETA_SKETCH_ESTIMATE(THETA_SKETCH_DIFF(DISTINCT_COUNT_RAW_THETA_SKETCH(int_col, ''), DISTINCT_COUNT_RAW_THETA_SKETCH(int_col, ''))), GET_THETA_SKETCH_ESTIMATE(THETA_SKETCH_UNION(DISTINCT_COUNT_RAW_THETA_SKETCH(int_col, ''), DISTINCT_COUNT_RAW_THETA_SKETCH(long_col, ''))), GET_THETA_SKETCH_ESTIMATE(THETA_SKETCH_INTERSECT(DISTINCT_COUNT_RAW_THETA_SKETCH(double_col, ''), DISTINCT_COUNT_RAW_THETA_SKETCH(float_col, 'nominalEntries=1000'))) FROM {tbl} GROUP BY bool_col",
+        "outputs": [[true, 59.666666666666664, 2, 0, 2, 3], [false, 88.5, 2, 0, 4, 3]]
       },
       {
         "sql": "select distinct_count_theta_sketch(int_col), distinctCountThetaSketch(long_col), DISTINCT_COUNT_THETA_SKETCH(float_col), DISTINCTCOUNTTHETASKETCH(double_col) FROM {tbl}",

